### ISSUE TITLE
Is there a formal error in the H_cc sector of newton_casscf.gen_g_hop?

### ISSUE THREAD

**Issue by MatthewRHermes** [pyscf/pyscf#701]
In differentiating with respect to an infinitesimal transformation of the CI vectors, we parameterize the energy as

<a href="https://www.codecogs.com/eqnedit.php?latex=E(x)=\langle0|e^{-\hat{P}}\hat{H}e^{\hat{P}}|0\rangle" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E(x)=\langle0|e^{-\hat{P}}\hat{H}e^{\hat{P}}|0\rangle" title="E(x)=\langle0|e^{-\hat{P}}\hat{H}e^{\hat{P}}|0\rangle" /></a>

with

<a href="https://www.codecogs.com/eqnedit.php?latex=\hat{P}\equiv\sum_{\vec{n}}|\vec{n}\rangle\langle0|-|0\rangle\langle\vec{n}|" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\hat{P}\equiv\sum_{\vec{n}}x_{\vec{n}}(|\vec{n}\rangle\langle0|-|0\rangle\langle\vec{n}|)" title="\hat{P}\equiv\sum_{\vec{n}}x_{\vec{n}}(|\vec{n}\rangle\langle0|-|0\rangle\langle\vec{n}|)" /></a>

The second derivative of the energy is

<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial^2&space;E}{\partial&space;x_{\vec{n}}\partial&space;x_{\vec{n}^\prime}}=\langle\vec{n}|\hat{Q}(\hat{H}-E_0)\hat{Q}|\vec{n}^\prime\rangle" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\partial^2&space;E}{\partial&space;x_{\vec{n}}\partial&space;x_{\vec{n}^\prime}}=\langle\vec{n}|\hat{Q}(\hat{H}-E_0)\hat{Q}|\vec{n}^\prime\rangle" title="\frac{\partial^2 E}{\partial x_{\vec{n}}\partial x_{\vec{n}^\prime}}=\langle\vec{n}|\hat{Q}(\hat{H}-E_0)\hat{Q}|\vec{n}^\prime\rangle" /></a>

with

<a href="https://www.codecogs.com/eqnedit.php?latex=\hat{Q}\equiv1-|0\rangle\langle0|" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\hat{Q}\equiv1-|0\rangle\langle0|" title="\hat{Q}\equiv1-|0\rangle\langle0|" /></a>
<a href="https://www.codecogs.com/eqnedit.php?latex=E_0\equiv\langle0|\hat{H}|0\rangle" target="_blank"><img src="https://latex.codecogs.com/gif.latex?E_0\equiv\langle0|\hat{H}|0\rangle" title="E_0\equiv\langle0|\hat{H}|0\rangle" /></a>

or, if we can forgive a bit of notational handwaving,

<a href="https://www.codecogs.com/eqnedit.php?latex=\partial^2&space;E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\partial^2&space;E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" title="\partial^2 E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" /></a>

But what's programmed in newton_casscf.py appears to be

<a href="https://www.codecogs.com/eqnedit.php?latex=\partial^2&space;E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-2|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-2(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\partial^2&space;E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-2|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-2(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" title="\partial^2 E|\vec{x}\rangle=(\hat{H}-E_0)|\vec{x}\rangle-2|0\rangle\langle0|(\hat{H}-E_0)|\vec{x}\rangle-2(\hat{H}-E_0)|0\rangle\langle0|\vec{x}\rangle" /></a>

in both hop and hdiag. I would understand a factor of 2 convention difference (EDIT: actually all three of my second-derivative expressions up there should have a +h.c. that I dropped), but this isn't a factor of 2, unless I've gotten confused about what (H-E0) is. In fact, it spoils intermediate normalization (we should have <0|x>=0 at all times; the component of P along the trial vector |0> is undefined by construction):

```
import numpy
from pyscf import gto, scf, lib, fci
from pyscf.mcscf import newton_casscf, CASSCF, addons

mol = gto.Mole()
mol.verbose = lib.logger.DEBUG
mol.output = '/dev/null'
mol.atom = [
    ['H', ( 5.,-1.    , 1.   )],
    ['H', ( 0.,-5.    ,-2.   )],
    ['H', ( 4.,-0.5   ,-3.   )],
    ['H', ( 0.,-4.5   ,-1.   )],
    ['H', ( 3.,-0.5   ,-0.   )],
    ['H', ( 0.,-3.    ,-1.   )],
    ['H', ( 2.,-2.5   , 0.   )],
    ['H', ( 1., 1.    , 3.   )],
]
mol.basis = 'sto-3g'
mol.build()
mf = scf.RHF(mol)
mf.max_cycle = 3
mf.kernel()
mc = newton_casscf.CASSCF(mf, 4, 4)
mc.fcisolver = fci.direct_spin1.FCI(mol)
mc.kernel()

numpy.random.seed(1)
mo = numpy.random.random(mf.mo_coeff.shape)
ci0 = numpy.random.random((6,6))
ci0/= numpy.linalg.norm(ci0)
gall, gop, hop, hdiag = newton_casscf.gen_g_hop(mc, mo, ci0, mc.ao2mo(mo))
x = numpy.random.random(gall.size)
hx = hop (x)
print ("|hx_ci| =",numpy.linalg.norm (hx[-36:]),"<0|hx> =", ci0.ravel ().dot (hx[-36:]))
```

yields

```
|hx_ci| = 73.30773999100957 <0|hx> = 4.13369273110702
```

I think this has minimal effects on the results of calculations because 1) the extract_rotation function forcibly normalizes the new output CI vector after every Newton-CG step, and the s10 business in the H_oc sector corrects for the effect of the spurious <0|x> component on the orbitals, so all this does is attenuate CI vector evolution almost like a trust radius, and 2) the coupled-perturbed solver for the state-average analytical gradient calculation that I added has a preconditioner that enforces <0|x> = 0 post hoc. But I wanted to ask about it to see if I was missing something, especially since changing it would make at least one test fail.

**Comment by verena-neufeld** [comment]
Hi @MatthewRHermes , thanks for having raised this issue. Since it has been such a while now, has it been resolved by now?

**Comment by MatthewRHermes** [comment]
I was wondering if there was some reason for this, but since no one has responded I decided to make the pull request to "fix" it.

**Comment by verena-neufeld** [comment]
Thanks @MatthewRHermes , I guess this issue is resolved now?

**Comment by MatthewRHermes** [comment]
Resolved by PR #1869
